### PR TITLE
[SPEC_V2] Fix Acclen drop when enabling DP Attention for Spec-Overlap

### LINKED ISSUE #16274: [Bug] SPEC_V2 EAGLE Speculative Decoding Accept Length Degraded with DP Attention on GB300
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When enabling DP (Data Parallel) attention (`--enable-dp-attention`) with EAGLE speculative decoding, the average accept length drops significantly compared to the same configuration without DP attention.

| Configuration | TP | DP | Chunked Prefill | Accept Length (avg) | Accept Rate (avg) |
|--------------|----|----|-----------------|---------------------|-------------------|
| TP4 only | 4 | 1 | 4096 | **2.3377** | 0.7795 |
| TP4 only | 4 | 1 | 8192 | **2.3541** | 0.7851 |
| TP4 only | 4 | 1 | 16384 | **2.3676** | 0.7897 |
| TP4 only | 4 | 1 | 32768 | **2.3711** | 0.7906 |
| **TP4 + DP4** | 4 | 4 | 4096 | **1.4931** | 0.4976 |
| **TP4 + DP4** | 4 | 4 | 8192 | **1.4819** | 0.4939 |

**The accept length drops from ~2.35 to ~1.48 (≈37% degradation) when enabling DP attention.**

### Reproduction Steps

1. **Without DP attention (baseline)**:
```

python -m sglang.launch_server \
    --model /path/to/DeepSeek-R1-0528-NVFP4-v2 \
    --tp-size 4 \
    --attention-backend trtllm_mla \
    --moe-runner-backend flashinfer_trtllm \
    --quantization modelopt_fp4 \
    --kv-cache-dtype fp8_e4m3 \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 2 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 3 \
    --chunked-prefill-size 8192 \
    --disable-radix-cache

```
**With DP attention (degraded)**:
```
python -m sglang.launch_server \
    --model /path/to/DeepSeek-R1-0528-NVFP4-v2 \
    --tp-size 4 \
    --dp-size 4 \
    --enable-dp-attention \
    --attention-backend trtllm_mla \
    --moe-runner-backend flashinfer_trtllm \
    --quantization modelopt_fp4 \
    --kv-cache-dtype fp8_e4m3 \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 2 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 3 \
    --chunked-prefill-size 8192 \
    --disable-radix-cache
```

### Environment

Python: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3: NVIDIA GB300
GPU 0,1,2,3 Compute Capability: 10.3
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.95.05
PyTorch: 2.9.1+cu130
sglang: 0.5.6.post2
sgl_kernel: 0.3.19
flashinfer_python: 0.5.3
flashinfer_cubin: 0.5.3
flashinfer_jit_cache: Module Not Found
triton: 3.5.1
transformers: 4.57.1
torchao: 0.9.0
numpy: 2.3.5
aiohttp: 3.13.2
fastapi: 0.124.4
hf_transfer: 0.1.9
huggingface_hub: 0.36.0
interegular: 0.3.3
modelscope: 1.33.0
orjson: 3.11.5
outlines: 0.1.11
packaging: 25.0
psutil: 7.1.3
pydantic: 2.12.5
python-multipart: 0.0.20
pyzmq: 27.1.0
uvicorn: 0.38.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.1.27
openai: 2.6.1
tiktoken: 0.12.0
anthropic: 0.75.0
litellm: Module Not Found
decord2: 2.0.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV16    NV16    NV16    NODE    NODE    SYS     SYS     SYS     SYS     0-71    0               2
GPU1    NV16     X      NV16    NV16    NODE    NODE    SYS     SYS     SYS     SYS     0-71    0               10
GPU2    NV16    NV16     X      NV16    SYS     SYS     NODE    NODE    NODE    NODE    72-143  1               18
GPU3    NV16    NV16    NV16     X      SYS     SYS     NODE    NODE    NODE    NODE    72-143  1               26
NIC0    NODE    NODE    SYS     SYS      X      NODE    SYS     SYS     SYS     SYS
NIC1    NODE    NODE    SYS     SYS     NODE     X      SYS     SYS     SYS     SYS
NIC2    SYS     SYS     NODE    NODE    SYS     SYS      X      NODE    NODE    NODE
NIC3    SYS     SYS     NODE    NODE    SYS     SYS     NODE     X      NODE    NODE
NIC4    SYS     SYS     NODE    NODE    SYS     SYS     NODE    NODE     X      PIX
NIC5    SYS     SYS     NODE    NODE    SYS     SYS     NODE    NODE    PIX      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3
  NIC4: mlx5_4
  NIC5: mlx5_5

ulimit soft: 1048576

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

Fix [#16274](https://github.com/sgl-project/sglang/issues/16274): SPEC_V2 EAGLE speculative decoding accept length degraded (~37% drop) when DP Attention is enabled.

## Modifications

Set `num_tokens_for_logprob_per_batch` to match `num_tokens_per_batch` in `_draft_extend_for_decode`.

**Root Cause**: `num_tokens_for_logprob_per_batch=1` caused `LogitsMetadata.compute_dp_attention_metadata` to calculate incorrect `dp_local_num_tokens` (4 instead of 12), resulting in MLP sync only processing partial tokens.
image

## Accuracy Tests

```
python -m sglang.test.run_eval --eval-name longbench_v2 --host 127.0.0.1 --port 30000 --model nvidia/DeepSeek-R1-0528-NVFP4-v2 --max-context-length 128000 --max-tokens 16384 --num-threads 16

Loaded 503 examples from LongBench-v2
Context length filter: None-128000 characters
ChatCompletionSampler initialized with self.system_message=None self.temperature=0.0 self.max_tokens=16384 self.reasoning_effort=None self.extra_body={}
  0%|          | 0/503 [00:00<?, ?it/s]Token indices sequence length is longer than the specified maximum sequence length for this model (135859 > 131072). Running this sequence through the model will result in indexing errors
100%|██████████| 503/503 [51:49<00:00,  6.18s/it]  
Total latency: 3109.123 s
Score: 0.588
```

**Before (accept_len ~1.5)**:
```
[DP2 TP2] Decode batch, accept len: 1.62, accept rate: 0.54
[DP1 TP1] Decode batch, accept len: 1.48, accept rate: 0.49
[DP3 TP3] Decode batch, accept len: 1.55, accept rate: 0.52
```

**After (accept_len ~2.4)**:
```
[DP1 TP1] Decode batch, accept len: 2.37, accept rate: 0.79
[DP2 TP2] Decode batch, accept len: 2.34, accept rate: 0.78
[DP3 TP3] Decode batch, accept len: 2.43, accept rate: 0.81
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

**Comment by Fridge003** [comment]
Just verified this fix on dpsk fp4 + dp + spec v2
```
SGLANG_ENABLE_SPEC_V2=1 \
python3 -m sglang.launch_server \
    --model-path nvidia/DeepSeek-R1-0528-FP4-v2 \
    --served-model-name dsr1 \
    --enable-dp-attention \
    --tp 4 \
    --dp 4 \
    --attention-backend trtllm_mla \
    --moe-runner-backend flashinfer_trtllm \
    --quantization modelopt_fp4 \
    --kv-cache-dtype fp8_e4m3 \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 3 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 4
    
python3 -m sglang.test.send_one
```

Before this fix:
```
+-------------+--------+------------+-----------------+
| Latency (s) | Tokens | Acc Length | Speed (token/s) |
+-------------+--------+------------+-----------------+
|    3.943    |  433   |   2.062    |     109.83      |
+-------------+--------+------------+-----------------+
```

After this fix:
```
+-------------+--------+------------+-----------------+
| Latency (s) | Tokens | Acc Length | Speed (token/s) |
+-------------+--------+------------+-----------------+
|    2.780    |  433   |   2.966    |     155.73      |
+-------------+--------+------------+-----------------+
```

**Review by Qiaolin-Yu — APPROVED** [review]
